### ISSUE TITLE
Trained Qwen 3 model seems to be broken.

### ISSUE THREAD

**Issue by ibnbd** [modelscope/ms-swift#4835]
I am trying to LoRA train a Qwen 3 8 billion model using the Supervised Fine-tuning dataset.
Here is an example dataset I am using 

```json
{
  "messages": [
    {
      "role": "system",
      "content": "You are an AI that manages calendar appointments.\n\nAllowed operations (via function calls):\n- check_availability(date: string, time: string, service_name: string)\n- schedule_appointment(date: string, time: string, service_name: string, client_name: string)"
    },
    {
      "role": "user",
      "content": "Hi, I'd like to book a haircut for tomorrow at 2 PM."
    },
    {
      "role": "assistant",
      "content": "<think>\n\n</think>\n\nLet me check this for you."
    },
    {
      "role": "tool_call",
      "content": "{\"name\": \"check_availability\", \"arguments\": {\"date\": \"2024-08-16\", \"time\": \"14:00\", \"service_name\": \"Haircut\"}}"
    },
    {
      "role": "tool_response",
      "content": "status=available"
    },
    {
      "role": "assistant",
      "content": "<think>\n\n</think>\n\nThat time is available. What is your name?"
    },
    {
      "role": "user",
      "content": "I'm Jordan."
    },
    {
      "role": "tool_call",
      "content": "{\"name\": \"schedule_appointment\", \"arguments\": {\"date\": \"2024-08-16\", \"time\": \"14:00\", \"service_name\": \"Haircut\", \"client_name\": \"Jordan\"}}"
    },
    {
      "role": "tool_response",
      "content": "result=Appointment successfully scheduled!|event_id=456-ABC"
    },
    {
      "role": "assistant",
      "content": "<think>\n\n</think>\n\nThanks, Jordan. You're all set for your haircut tomorrow at 2:00 PM."
    }
  ]
}
```

Here is the 'TrainArguments'  I am using 

```
    sft_args = TrainArguments(
        # Model Configuration
        model='Qwen/Qwen3-8B',
        
        # Dataset Configuration
        dataset=['data/swift_format.jsonl'],
        
        # CRITICAL: For Qwen3 reasoning model, must use ignore_empty_think
        # This preserves reasoning capabilities when <think> tags are empty
        loss_scale='ignore_empty_think',
        
        # Qwen-specific agent template for tool handling
        agent_template='qwen_en',
        
        # Training Configuration - Focus on data quality over aggressive training
        train_type='lora',
        num_train_epochs=3,  # Standard epochs, rely on data quality
        max_length=4096,
        learning_rate=5e-6,  # Very conservative LR to prevent overfitting to tool patterns
        per_device_train_batch_size=2,
        gradient_accumulation_steps=16,
        warmup_ratio=0.2,   # Longer warmup for stability
        weight_decay=0.1,   # Higher weight decay to prevent tool overfitting
        
        # Optimization Settings
        torch_dtype='bfloat16',
        gradient_checkpointing=False,
        
        # LoRA Configuration - Conservative to prevent overfitting
        target_modules='all-linear',
        lora_rank=64,       # Lower rank to prevent overfitting to tool patterns
        lora_alpha=128,     # Balanced alpha
        lora_dropout=0.15,  # Higher dropout to improve generalization
        
        # Output Configuration
        output_dir='./output/qwen3-8b-h200-balanced',
        
        # Logging and Saving
        save_steps=250,     # More frequent saves to catch best model
        logging_steps=10,   # More frequent logging
        save_strategy='steps',
        
        # Performance Optimizations
        dataloader_num_workers=8,
        group_by_length=True,
        max_grad_norm=0.3,  # Lower gradient clipping for stability
    )
```

And here is the model merging I am using 

```
swift export \
    --adapters "$CHECKPOINT_PATH" \
    --merge_lora true
```

I have two issues with the training and the merge.

- The model loses reasoning capabilities.
- The model is too aggressive with tool calls, even though half of my dataset doesn't have tool calls.

I followed the docs but not sure if I'm missing something.
Any help will be appreciated.

**Comment by Jintao-Huang** [comment]
Do you have a dataset with mixed general reasoning?

https://swift.readthedocs.io/en/latest/BestPractices/Qwen3-Best-Practice.html

**Comment by leipan797** [comment]
I encountered the same issue. The swift seems to delete the ```<think>\n\n</think>\n\n``` pattern once trigger on the ignore_empty_think.

**Comment by sahilqure** [comment]
dont use qwen agent template, keep it to default and it will use hermes template. That will solve the issue

**Comment by berserkr** [comment]
@sahilqure what loss should we use? Bc I see loss scaling for tool calling and loss for thinking. Thank you :)

**Comment by dipta007** [comment]
I am facing the same issue, any solution here?

**Comment by gnohgnailoug** [comment]
> I encountered the same issue. The swift seems to delete the `<think>\n\n</think>\n\n` pattern once trigger on the ignore_empty_think.

I encounter this issue too.....

**Comment by z0o0ey** [comment]
please retry with latest swift and vllm, feel free to create a new issue if it the problem persists.
